File: server/sound/YouTube.ts
import sr, { Video } from "youtube-sr";
import { Downloadable } from "../struct";
import { join } from "path";
import fs from "fs";
import { exec } from "youtube-dl-exec";
import { Media } from "../server";
import { sanitizeFileName } from "../MediaManager";
import { DateTime } from "luxon";

export async function searchYoutube(query: string): Promise<Downloadable[]> {
  try {
    const results =
      (await sr.search(query, {
        type: "video",
      })) || [];
    return results.map(constructVideoFromYouTube);
  } catch (err) {
    console.error(
      `Error searching youtube: ${err}\nQuery: '${query}'\n${err.stack}`
    );
    return [];
  }
}

export async function downloadYoutube(
  url: string
): Promise<{ err: string } | string> {
  return new Promise(async (res) => {
    try {
      const path = join(Media.tempdir(), `${sanitizeFileName(url)}.mp4`);
      const file = fs.createWriteStream(path);
      const stream = exec(url, {
        format: "bestaudio",
        output: "-",
      });
      stream.stdout.pipe(file);
      file.on("error", (err) => {
        res({ err: String(err) });
      });
      file.on("finish", () => {
        res(path);
      });
    } catch (err) {
      res({ err });
    }
  });
}

export function constructVideoFromYouTube(vid: Video): Downloadable {
  if (!Number(vid.uploadedAt[0]))
    vid.uploadedAt = vid.uploadedAt.split(" ").slice(1).join(" ");
  return {
    title: vid.title || "",
    duration: vid.duration || 0,
    uploadedAt: vid.uploadedAt
      ? DateTime.now()
          .minus({
            [vid.uploadedAt.split(" ")[1]]: Number(
              vid.uploadedAt.split(" ")[0]
            ),
          })
          .minus({ minute: 1 }) // fixes the time
          .toMillis()
      : Date.now(),
    views: vid.views || 0,
    thumbnail: vid.thumbnail?.url || "",
    author: {
      name: vid.channel?.name || "",
      icon: vid.channel?.iconURL() || "",
      url: vid.channel?.url || "",
    },
    url: vid.url || "",
    embed: vid.embedURL || "",
  };
}

File: server/sound/youtube.go
package sound

import (
	"errors"
	"io"
	"log"
	"os"
	"path"
	"strings"
	"time"

	"github.com/itzTheMeow/YTMusic/types"
	"github.com/levigross/grequests"
	"github.com/oklog/ulid/v2"
)

type pipedAPISearchResponse struct {
	URL        string `json:"url"`  // is "/watch?v=VIDEO_ID"
	Type       string `json:"type"` // enforce "stream"
	Title      string `json:"title"`
	Thumbnail  string `json:"thumbnail"`
	AuthorName string `json:"uploaderName"`
	AuthorURL  string `json:"uploaderUrl"` // is "/channel/CHANNEL_ID"
	AuthorIcon string `json:"uploaderAvatar"`
	Duration   int    `json:"duration"`
	Views      int    `json:"views"`
	UploadedAt int    `json:"uploaded"`
	IsShort    bool   `json:"isShort"` // enforce "false"
	// ignore uploadedDate, shortDescription, uploaderVerified
}

func SearchYoutube(query string) []types.Downloadable {
	res, err := grequests.Get("https://pipedapi.tokhmi.xyz/search", &grequests.RequestOptions{
		Params: map[string]string{
			"q":      query,
			"filter": "videos",
		},
	})
	if err != nil {
		log.Println("Error fetching youtube search results.", err)
		return make([]types.Downloadable, 0)
	}
	var body struct {
		Items []pipedAPISearchResponse `json:"items"`
	}
	res.JSON(&body)
	list := make([]types.Downloadable, 0)
	for _, vid := range body.Items {
		if vid.Type == "stream" && !vid.IsShort {
			list = append(list, ConstructDownloadableFromYouTube(vid))
		}
	}
	return list
}

func ConstructDownloadableFromYouTube(vid pipedAPISearchResponse) types.Downloadable {
	if vid.Duration == -1 {
		vid.Duration = 0
	}
	if vid.UploadedAt == -1 {
		vid.UploadedAt = int(time.Now().UnixMilli())
	}
	return types.Downloadable{
		Title:      vid.Title,
		Duration:   vid.Duration * 1000,
		UploadedAt: vid.UploadedAt,
		Views:      vid.Views,
		Thumbnail:  vid.Thumbnail,
		Author: types.DownloadableAuthor{
			Name: vid.AuthorName,
			Icon: vid.AuthorIcon,
			URL:  "https://youtube.com" + vid.AuthorURL,
		},
		URL:   "https://youtube.com" + vid.URL,
		Embed: "https://youtube.com" + strings.Replace(vid.URL, "/watch?v=", "/embed/", 1),
	}
}

func DownloadYouTube(url string) (*string, error) {
	filepath := path.Join(os.TempDir(), ulid.Make().String())

	result, err := DownloadURL(url)
	if err != nil {
		return nil, err
	}
	file, err := os.Create(filepath)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	written, err := io.Copy(file, result)
	if err != nil {
		return nil, err
	}
	if written == 0 {
		return nil, errors.New("Failed to download video.")
	}
	return &filepath, nil
}
